pkglistgen: Support working on git branches

adrianschroeter opened the pull request:
Following scmsync syntax where a branch can be specified behind the URL fragment hash

Vogtinator commented on pkglistgen/tool.py:
```diff
@@ -703,8 +703,13 @@ def update_and_solve_target(
+                git_url = split[0]
+                args = ['git', 'clone', '--recurse-submodules', git_url, cache_dir]
+                if len(split) == 2:
+                    args.append('--branch')
```
`--single-branch` as well (unconditionally)?

adrianschroeter replied:
we can do so ... added another commit for that as it is independend of the branch selection.

Vogtinator commented on pkglistgen/tool.py:
```diff
@@ -703,8 +703,13 @@ def update_and_solve_target(
                     shutil.rmtree(cache_dir)
-                logging.debug(subprocess.check_output(
-                    ['git', 'clone', '--recurse-submodules', git_url, cache_dir], encoding='utf-8'))
+                split = git_url.split('#')
+                git_url = split[0]
+                args = ['git', 'clone', '--single-branch', '--recurse-submodules', git_url, cache_dir]
+                if len(split) == 2:
+                    args.append('--branch')
+                    args.append(split[1])
+                logging.debug(subprocess.check_output(args, encoding='utf-8'))
```
```suggestion
                git_url_base, *fragment = git_url.split('#')
                args = ['git', 'clone', '--single-branch', '--recurse-submodules', git_url_base, cache_dir]
                if fragment:
                    args += ['--branch', fragment[0]]
                logging.debug(subprocess.check_output(args, encoding='utf-8'))
```

Vogtinator commented:
Should it also handle if the URL (any part) changes while the cache dir already exists?
Currently it'll pull and push to the previous location.

adrianschroeter commented:
yes, but that can get complex as also the default branch can change server side...

maybe we should do that with a second pull request, to avoid blocking Marco's request.

Vogtinator commented:
> yes, but that can get complex as also the default branch can change server side...

That only impacts the initial clone though, so I don't see how that would be complex?

If it's ok to discard the cache dir on mismatches, then `git remote get-url origin` and `git branch --show-current` could be used to check for mismatches.

Vogtinator commented:
Without support for cache invalidation it would currently keep pulling and pushing the wrong URL

adrianschroeter commented:
I added a 90% switch solution here, but it does not cover the situation when no branch is specified, but default branch switches server side.

I reverted the --single-branch clone option therefore to be able to switch later on.

Vogtinator commented on pkglistgen/tool.py:
```diff
@@ -696,17 +696,24 @@ def update_and_solve_target(
+                if fragment:
+                    # FIXME: this is not takeing account default branches, which may change server side
+                    logging.debug(subprocess.check_output(
+                        ['git', 'switch', fragment], cwd=cache_dir, encoding='utf-8'))
```
`git switch` is experimental and might not be available on ibs-slsa-botmaster

lethliel replied:
the git on botmaster knows git switch

lethliel replied:
also git on ibs-slsa-botmaster knows it

gleidi-suse commented:
@adrianschroeter @Vogtinator we need this for SLFO stagings. Is there any action to be taken or can this be merged?

adrianschroeter commented:
it is fine from my side.

Please speak up if something needs to be changed.
